docker/Dockerfile.build

```diff
@@ -1,56 +1,101 @@
-# Builder image
-FROM rust:buster as builder
-
-RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
-
-RUN mkdir /zebra
-WORKDIR /zebra
-
-ARG SHORT_SHA
-ENV SHORT_SHA $SHORT_SHA
+# This steps implement cargo-chef for docker layer caching
+# We are using four stages:
+# - chef: installs cargo-chef
+# - planner: computes the recipe file
+# - builder: caches our dependencies and builds the binary
+# - tester: builds and run tests
+# - runtime: is our runtime environment
+FROM rust:bullseye as chef
+RUN cargo install cargo-chef --locked
+WORKDIR /app
+
+FROM chef AS planner
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+FROM chef AS builder
+COPY --from=planner /app/recipe.json recipe.json
+
+# Install zebra build deps
+RUN apt-get -qq update && \
+    apt-get -qq install -y --no-install-recommends \
+    llvm \
+    libclang-dev \
+    clang \
+    ca-certificates \
+    ; \
+    rm -rf /var/lib/apt/lists/* /tmp/*
+
+# Install google OS Config agent
+RUN apt-get -qq update && \
+    apt-get -qq install -y --no-install-recommends \
+    curl \
+    lsb-release \
+    ; \
+    echo "deb http://packages.cloud.google.com/apt google-compute-engine-$(lsb_release -cs)-stable main" > /etc/apt/sources.list.d/google-compute-engine.list && \
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add - && \
+    apt-get -qq update  && \
+    apt-get -qq install -y --no-install-recommends google-osconfig-agent && \
+    rm -rf /var/lib/apt/lists/* /tmp/*
+
+# Optimize builds. In particular, regenerate-stateful-test-disks.yml was reaching the
+# GitHub Actions time limit (6 hours), so we needed to make it faster.
+ENV RUSTFLAGS -O
+ENV CARGO_HOME /app/.cargo/
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
+
+ARG RUST_BACKTRACE=1
+ENV RUST_BACKTRACE ${RUST_BACKTRACE}
 
-ENV RUST_BACKTRACE full
-ENV CARGO_HOME /zebra/.cargo/
+COPY . .
+# Pre-download Zcash Sprout and Sapling parameters
+RUN cargo run --locked --release --features enable-sentry --bin zebrad download
+# Build zebra
+RUN cargo build --locked --release --features enable-sentry --bin zebrad
 
-RUN rustc -V; cargo -V; rustup -V
+FROM builder AS tester
+# Skip IPv6 tests by default, as some CI environment don't have IPv6 available
+ARG ZEBRA_SKIP_IPV6_TESTS=1
+ENV ZEBRA_SKIP_IPV6_TESTS ${ZEBRA_SKIP_IPV6_TESTS}
 
-COPY . .
+RUN cargo test --locked --release --features enable-sentry --workspace --no-run
 
-RUN cd zebrad/; cargo build --release --features enable-sentry
+CMD ["cargo" "test" "--locked" "--release" "--features" "enable-sentry" "--workspace"]
```

Comment:
👍

Not sure if we need `--enable-sentry` here

Reply:
This is mainly to avoid rebuilding. It's not needed, but it's preferable to have it.

Reply:
OK sure